node/network/approval-distribution/src/lib.rs

```diff
@@ -233,31 +227,31 @@ enum Resend {
 ///
 /// It tracks metadata about our view of the unfinalized chain,
 /// which assignments and approvals we have seen, and our peers' views.
-#[derive(Default)]
+#[derive(Clone, Default)]
 struct State {
 	/// These two fields are used in conjunction to construct a view over the unfinalized chain.
-	blocks_by_number: BTreeMap<BlockNumber, Vec<Hash>>,
-	blocks: HashMap<Hash, BlockEntry>,
+	blocks_by_number: Arc<Mutex<BTreeMap<BlockNumber, Vec<Hash>>>>,
```

Comment:
This many mutexes scares me, frankly. Lock order needs to be well-defined to avoid deadlocks.

Reply:
We should just `Arc` the whole state, not `Arc` everything in the state.

Reply:
You are right. I wanted to go for more granular locking as I did not know that we can't await with holding the mutex, but now that this is no longer happening, I can just lock the whole state.

Reply:
That's actually not what I meant.

```rust
state: Arc<State>

struct State {
    a: Mutex<A>,
    b: Mutex<B>,
}
```

vs 

```rust
state: State

struct State {
    a: Arc<Mutex<A>>,
    b: Arc<Mutex<B>>,
}
```

vs 

```rust
state: Arc<Mutex<State>

struct State {
    a: A,
    b: B,
}
```

are all different.

Reply:
Yup 🤯